File: CP/TCS-Codevita/CodeVita2017-master/Round1/FindRank.java
import java.util.Scanner;
import java.math.BigInteger;

public class FindRank
{
    private static long fact(long n)
    {
        return (n <= 1) ? 1 : n * fact(n - 1);
    }

    private static int findSmallerInRight(String str, int low, int high)
    {
        int countRight = 0;

        for (int i = low + 1; i <= high; i++)
            if (str.charAt(i) < str.charAt(low))
                countRight++;

        return countRight;
    }

    private static long findRank(String str)
    {
        int len = str.length();
        long mul = fact(len);

        long rank = 1;

        int countRight;

        for (int i = 0; i < len; i++)
        {
            mul /= len - i;

            countRight = findSmallerInRight(str, i, len - 1);
            rank += countRight * mul;
        }

        return rank;
    }

    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);

        int n = Integer.parseInt(sc.nextLine().trim());
        BigInteger r = BigInteger.valueOf(1l);

        for (int i = 0; i < n; i++)
        {
            String line = sc.nextLine().trim();
            r = r.multiply(BigInteger.valueOf(findRank(line)));
        }

        r = r.mod(BigInteger.valueOf(23456));

        System.out.println(r);

        sc.close();
    }
}

File: CP/TCS-Codevita/CodeVita2017-master/Round1/findrank.c
#include <stdio.h>
#include <string.h>

long fact(long n)
{
	return (n <= 1) ? 1 : n * fact(n - 1);
}

// A utility function to count smaller characters on right
// of str[low]
int findSmallerInRight(char* str, int low, int high)
{
	int countRight = 0, i;

	for (i = low + 1; i <= high; i++)
		if (str[i] < str[low])
			countRight++;

	return countRight;
}

// A function to find rank of a string in all permutations
// of characters
long findRank (char* str)
{
	int len = strlen(str);
	long mul = fact(len);
	long rank = 1;
	int countRight;
	int i;

	for (i = 0; i < len; i++)
	{
		mul /= len - i;

		// count number of chars smaller than str[i]
		// fron str[i + 1] to str[len - 1]
		countRight = findSmallerInRight(str, i, len - 1);

		rank += countRight * mul ;
	}

	return rank;
}

int main()
{
    int n, i, j, k;
    char a[11];
    long long r = 1;

    scanf("%d", &n);

    for (i = 0; i < n; i++)
    {
        a[10] = '\0';

        for (j = 0; j < 10; j++)
        {
            scanf("%1d", &k);
            a[j] = k + '0';
        }

        r *= findRank(a);
    }

    r %= 23456;

    printf("%ll\n", r);

	return 0;
}
